### ISSUE TITLE
oauth token race condition?

### ISSUE THREAD

**Issue by riptusk331** [O365/python-o365#350]
I have run into a fairly interesting race condition problem using oauth tokens stored on the filesystem. At least it has the hallmarks of a race condition...

I use O365 in a (Docker) containerized application I've developed that manages file encryption & SFTP transfers for my company. My application runs various download & upload jobs daily via **cron**, and sends out status e-mails when the jobs have completed. I have wrapped O365 in a separate module I created that takes care of all its initialization and allows my other modules to easily send preset e-mails depending on the job. Each job calls its own instance of this O365 wrapper.

I have 2 jobs that run at the exact same time in the morning. The jobs take **_slightly_** different amounts of time to complete, on the order of milliseconds. When they're done, they both call the O365 wrapper, which starts the process of initializing O365 and sending the desired e-mails. At the time of these jobs, the last token was used over 60 min prior, so O365 has to fetch new oAuth tokens thru Graph.

Here's where things get funky. Since the calls to O365 come within a few milliseconds of one another, it's not enough time for the first call to fetch and save down a new token **and** have the other call see this token before it goes and does the same thing. So both O365 calls think they need to get new tokens...which they do (successfully). Both jobs then send a POST request to the graph sendMail endpoint with my email content...and after about 20 seconds I get a 503 service unavailable error on both requests.

To make this a bit easier to understand - here is a very rough, simplified & scrubbed version of my logs, to indicate what's happening.

```
2019-11-06 07:00:01,236 - job1 - INFO - Job 1 Finishes
2019-11-06 07:00:01,238 - O365.connection:_internal_request:618 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
2019-11-06 07:00:01,238 - O365.connection:_internal_request:619 - INFO - Request parameters: {...job 1 graph json...}
2019-11-06 07:00:01,238 - O365.connection:_internal_request:634 - INFO - Oauth Token is expired, fetching a new token
2019-11-06 07:00:01,246 - job2 - INFO - Job 2 Finishes
2019-11-06 07:00:01,248 - O365.connection:_internal_request:618 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
2019-11-06 07:00:01,248 - O365.connection:_internal_request:619 - INFO - Request parameters: {...job 2 graph json...}
2019-11-06 07:00:01,248 - O365.connection:_internal_request:634 - INFO - Oauth Token is expired, fetching a new token
2019-11-06 07:00:01,669 - O365.connection:_internal_request:637 - INFO - New oauth token fetched
2019-11-06 07:00:01,669 - O365.connection:_internal_request:618 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
2019-11-06 07:00:01,669 - O365.connection:_internal_request:619 - INFO - Request parameters: {...job 1 graph json...}
2019-11-06 07:00:01,734 - O365.connection:_internal_request:637 - INFO - New oauth token fetched
2019-11-06 07:00:01,734 - O365.connection:_internal_request:618 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
2019-11-06 07:00:01,735 - O365.connection:_internal_request:619 - INFO - Request parameters: {...job 2 graph json...}
2019-11-06 07:00:24,354 - O365.connection:_internal_request:662 - DEBUG - Server Error: 503 Server Error: Service Unavailable for url: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
2019-11-06 07:00:24,582 - O365.connection:_internal_request:662 - DEBUG - Server Error: 503 Server Error: Service Unavailable for url: https://graph.microsoft.com/v1.0/users/myemail@myemail.com/sendMail
```

I suspect the issue is coming from the first job sending a POST request to send the email while the graph server is still handling the second job's request for a new token. The 503 code on the MS documentation indicates the service is not available OR busy. But I'm not positive. MS server is definitely not down. I wanted to see if you had encountered this or something similar before, and had any suggestions.

The easiest solution is to just have the job's run a minute apart. Gives enough time to fetch and read new tokens. Or I thought of maybe using a token file name unique to the job? Currently, I use the same token file for everything.

Any input would be appreciated. Thanks!

**Comment by janscas** [comment]
Hi, this is a **great issue**!!

The case here is that each O365 `Connection` instance is not aware of the existence of the other instances. So all of them will try to refresh the token (even at once if you run them in parallel). 

Yes, there is a race condition. But the server responding with a 503 is odd. It should be a 4XX status code.

Another thing here is that the instances should know when a token is updated. Because there should be only one call to refresh the token. All the other calls even if they succeed are not necessary. Image you had 1.000 jobs. That will be 1.000 calls to refresh the token and this is a tremendous unnecessary effort.

But this is not easy to solve.

### In the next lines I'm just thinking out loud...

Maybe in the event of a refresh token action we can make each `Connection` instance to always ask the TokenBackend if it should initiate the token refresh action (something like `if self.con.token_backend.should_refresh_token():`).

The TokenBackend can then check some condition to answer. This is completely up to the TokenBackends. Some TokenBackends can be programmed to always answer "yes". Others will have to check more things.

- If the TokenBackend is a Database this is easily achieved. For example with a flag stored along the token info that will be updated when the new token gets saved and can be checked with a transaccional call. If the flag is something like "ready" to refresh then updates the flag to "updating". When the refresh is complete the TokenBackend should save the token and then update the flag state to "ready" again.

- If the TokenBackend is the `FileSystemTokenBackend` I really don't know how to achieve this. The file system does not have transactional atomic operations. Maybe with the `FileSystemTokenBackend` the answer should be always True... but then the same problem you notice will arise and it will have to be solved by other means like you did.

The answer to the `should_refresh_token` call can be:

- True: then the `Connection` will execute the refresh and store the token also updating the flag from "updating" to something like "ready" again.

- False: then the `Connection` will know it has to wait until the answer is True again. So it will incrementally (with a tight backoff factor) ask again until the answer is True. But in this case instead of refreshing the token, it should re-load the token from the backend to get the new access token.

These are just some thoughs on this....
I don't have a clear answer on how to solve this problematic.

What do you think?

**Comment by riptusk331** [comment]
I like your `should_refresh_token()` solution. I have not tried to store the token in a DB before, but conceptually I think in the case of something like a database with atomic operations, this solution would work great. A simple 'updating/ready' flag, as you suggest, is clean and simple. I don't think this refresh check would add much overhead either.

The `FileSystemTokenBackend` is a bit more tricky. I'll need to think about this and toy around with the library when I have some free time. I'll report back and if we can come up with something neat, I'll submit a PR. A fun problem to solve for sure.

For now, I am just going to separate the jobs by a minute as a tactical fix to my issue.

**Comment by janscas** [comment]
I was thinking about this.
Maybe you should refresh the token manually.

You can call `account.con.refresh_token()` at any point to refresh the access token.

Maybe you can call this manually before starting the x jobs that send emails, so all of them can use a new fresh access token. This will avoid all the calls to refresh tokens.

**Comment by riptusk331** [comment]
I'll give that a try. My concern there is that I'd still be making concurrent requests for new tokens if 2 jobs start at the same time, which seems to be causing my issues here. But it's worth a shot. Or are you saying to schedule a single token refresh on its own before starting the jobs at the same time? In that case, I think that my original solution of starting one job a minute before the other in effect does that, since whichever runs first will trigger the auto-refresh.

I've been doing some research on how to make the `FileSystemTokenBackend` refresh atomic, and I think there is hope. Python has the very interesting [fcntl](https://docs.python.org/3.8/library/fcntl.html) built-in package that allows for file locking...it's essential a wrapper around the UNIX `fcntl` utility. The downside is that it only works on UNIX-like systems. There is another nice package I have been testing called [portalocker](https://github.com/WoLpH/portalocker)...it is essentially a wrapper around `fcntl` and `win32con`, which expands the locking ability to Windows based systems.

I'm going to play around with these and see if I can implement them in a nice way. Just thinking out loud...my thoughts are that we could apply a exclusive lock (`LOCK_EX`) on the `token.txt` file when it is first read, and release the lock only when a newly refreshed token is written or it's determined a new one isn't needed. Then we just put in some kind of retry logic on the token reader...if it tries to check for and read the token and the file is locked, sleep for a few seconds and try again. Quit after 5 unsuccessful retries as not to hang.

So very crudely:
**1.** job 1 finishes & calls/initializes o365
**2.** o365 instance 1 opens & reads token file - applies an exclusive lock
**3.** o365 instance 1 determines token is expired....go get a new one from ms-graph server
**4.** job 2 finishes & calls/initalizes o365
**5.** o365 instance 2 tries to open & read token file - is blocked due to lock
**6.** o365 instance 2 goes to sleep for a few seconds
**7.** o365 instance 1 receives new token & writes it - lock is released
**8.** o365 instance 1 proceeds as normal
**9.** o365 instance 2 wakes up, tries to open file again - succeeds this time, and applies its own lock
**10.** o365 instance 2 determines token is current - lock is released
**11.** o365 instance 2 proceeds as normal..........

I am just spitballing here...no idea how feasible this may be to implement...but I would like to give it a stab. What are your thoughts?

**Comment by janscas** [comment]
> I'll give that a try. My concern there is that I'd still be making concurrent requests for new tokens if 2 jobs start at the same time, which seems to be causing my issues here. But it's worth a shot. Or are you saying to schedule a single token refresh on its own before starting the jobs at the same time? In that case, I think that my original solution of starting one job a minute before the other in effect does that, since whichever runs first will trigger the auto-refresh.

That is only true if the first job stores the new refreshed token before the other loads it. Start the second and following jobs 5 seconds after the first one that will request the token and you should see this behaviour.



> I am just spitballing here...no idea how feasible this may be to implement...but I would like to give it a stab. What are your thoughts?

Great! I'll like to see if this file-locking works as expected.
I've been working on this and I've already implemented an example of this with a google firestore DB. I'll reference the commit when it's ready.

**Comment by janscas** [comment]
see 295cb49cc363781fd87ee7dda5692c9ebc855e8c

**Comment by riptusk331** [comment]
Very nice! I think this is a great solution for the database store.

I've made good progress on a solution for the `FileSystemTokenBackend`. It will require more significant testing, but it's working quite nicely as I've tested it so far. I'm going to to submit a PR on another branch so you can take a look, but it pretty much functions as I described in our brainstorming above. And I tried to integrate the design with your `should_refresh_token` idea and recent commit implementing this into the `BaseTokenBackend`.

I had to add 2 new properties to the `Token` class to give us the **_access_** expiration vs the refresh expiration (which the other properties give by checking for a refresh token and adding 90 days). I use checking these properties as "should refresh" conditions for a slightly modified `FileSystemTokenBackend` that implements the `should_refresh_token` method.

The token backend by itself cannot entirely handle the file blocking since the refresh of the token is triggered by the `Connection` class. But it can at least be aware of its state of being blocked or not, through a newly added `fs_wait` attribute I added. This serves as the conditions to check in the `should_refresh_token` implementation for the `FileSystemTokenBackend`. If we're not blocked, you should refresh the token. That means we're the first to it. If we are blocked, go to sleep, wake up, and check the token again. If it's not expired and we're unlocked, that means another instance got a fresh token, and we can just use that one (and not make another token request).

At the `Connection` level, I added a  new `_fs_token_lock` method that basically locks the token file during the refresh process. The method also checks if the token file is locked already (by another Connection instance trying to refresh). If it finds the file is locked,  it sets the token backend's `fs_wait` attribute to True, and puts the thread to sleep for 2 seconds and tries to refresh again.

To call this method, I modified the behavior of `_internal_request` in the catching of a `TokenExpiredError`. It will check if the `token_backend` is a `FileSystemTokenBackend`, and if it is, calls my method. If not, proceed as normal...since locking at the request level is only needed by a file system stored token.

Take a look and let me know what you think! I hobbled this together over a few days, and it definitely requires testing...but I think it's a nice solution to our token access concurrency problem on the _**file**_ end.

**Comment by riptusk331** [comment]
See #356

**Comment by riptusk331** [comment]
Here is a dump of my logs - I've only tested this on a Mac...but you can see the instance encounter a locked file, go to sleep, wake up, and try again. When it finds that the token has been changed by another instance, it loads and uses that one. It's pretty cool. I am going to test this further in Docker, but maybe you could try on windows? Testing it is pretty easy. I use cron to simultaneously call 2 test scripts. Then I use `time.sleep()` in python to stagger them milliseconds apart (to replicate how where I was encountering issues in the first place. I found the file lock to be effective down to 1 millisecond, which is pretty nice (meaning the file was locked by 1 instance, and the other instance tried accessing a millisecond later and was blocked).

I found a few bugs & changes I had to make after I initially submitted my pull request. You'll see my pull request has 4 commits.

```
2019-11-10 18:56:00,586 - root:<module>:13 - INFO - JOB 1 sleeping for 0.005 seconds
2019-11-10 18:56:00,586 - root:<module>:13 - INFO - JOB 2 sleeping for 0.01 seconds
2019-11-10 18:56:00,683 - O365.connection:_internal_request:721 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:00,683 - O365.connection:_internal_request:722 - INFO - Request parameters: {'data'}
2019-11-10 18:56:00,683 - requests_oauthlib.oauth2_session:request:386 - DEBUG - Invoking 0 protected resource request hooks.
2019-11-10 18:56:00,683 - requests_oauthlib.oauth2_session:request:392 - DEBUG - Adding token {'token stuff'} to request.
2019-11-10 18:56:00,696 - O365.connection:_internal_request:721 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:00,696 - O365.connection:_internal_request:722 - INFO - Request parameters: {'params'}
2019-11-10 18:56:00,696 - requests_oauthlib.oauth2_session:request:386 - DEBUG - Invoking 0 protected resource request hooks.
2019-11-10 18:56:00,696 - requests_oauthlib.oauth2_session:request:392 - DEBUG - Adding token {'token_stuff'} to request.
**2019-11-10 18:56:00,724 - O365.connection:_internal_request:740 - INFO - Oauth Token is expired, fetching a new token**
**2019-11-10 18:56:00,724 - O365.connection:_fs_token_lock:666 - DEBUG - Locked oauth token file**
2019-11-10 18:56:00,724 - requests_oauthlib.oauth2_session:refresh_token:347 - DEBUG - Adding auto refresh key word arguments {}.
2019-11-10 18:56:00,724 - requests_oauthlib.oauth2_session:refresh_token:352 - DEBUG - Prepared refresh token request body grant_type=refresh_token&
2019-11-10 18:56:00,725 - requests_oauthlib.oauth2_session:request:421 - DEBUG - Requesting url https://login.microsoftonline.com/common/oauth2/v2.0/token using method POST.
2019-11-10 18:56:00,725 - requests_oauthlib.oauth2_session:request:422 - DEBUG - Supplying headers {'headers'}
2019-11-10 18:56:00,725 - requests_oauthlib.oauth2_session:request:423 - DEBUG - Passing through key word arguments {'args'}
2019-11-10 18:56:00,736 - O365.connection:_internal_request:740 - INFO - Oauth Token is expired, fetching a new token
**2019-11-10 18:56:00,736 - O365.connection:_fs_token_lock:676 - WARNING - Oauth file locked. Sleeping for 2 seconds...retrying 4 more times.**
2019-11-10 18:56:00,758 - urllib3.connectionpool:_new_conn:955 - DEBUG - Starting new HTTPS connection (1): login.microsoftonline.com:443
2019-11-10 18:56:01,060 - urllib3.connectionpool:_make_request:428 - DEBUG - https://login.microsoftonline.com:443 "POST /common/oauth2/v2.0/token HTTP/1.1" 200 3427
2019-11-10 18:56:01,061 - requests_oauthlib.oauth2_session:refresh_token:364 - DEBUG - Request to refresh token completed with status 200.
2019-11-10 18:56:01,061 - requests_oauthlib.oauth2_session:refresh_token:366 - DEBUG - Response headers were {'response'}.
2019-11-10 18:56:01,061 - requests_oauthlib.oauth2_session:refresh_token:368 - DEBUG - Invoking 0 token response hooks.
2019-11-10 18:56:01,062 - O365.connection:_fs_token_lock:670 - INFO - New oauth token fetched
2019-11-10 18:56:01,062 - O365.connection:_fs_token_lock:672 - DEBUG - Unlocked oauth token file
2019-11-10 18:56:01,062 - O365.connection:_internal_request:721 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:01,062 - O365.connection:_internal_request:722 - INFO - Request parameters: {'params'}
2019-11-10 18:56:01,062 - requests_oauthlib.oauth2_session:request:386 - DEBUG - Invoking 0 protected resource request hooks.
2019-11-10 18:56:01,062 - requests_oauthlib.oauth2_session:request:392 - DEBUG - Adding token {'token'} to request.
2019-11-10 18:56:01,062 - requests_oauthlib.oauth2_session:request:421 - DEBUG - Requesting url https://graph.microsoft.com/v1.0/users/<redact>/sendMail using method post.
2019-11-10 18:56:01,062 - requests_oauthlib.oauth2_session:request:422 - DEBUG - Supplying headers {'headers'} and data {'data'}
2019-11-10 18:56:01,062 - requests_oauthlib.oauth2_session:request:423 - DEBUG - Passing through key word arguments {}.
2019-11-10 18:56:01,064 - urllib3.connectionpool:_new_conn:955 - DEBUG - Starting new HTTPS connection (1): graph.microsoft.com:443
2019-11-10 18:56:01,519 - urllib3.connectionpool:_make_request:428 - DEBUG - https://graph.microsoft.com:443 "POST /v1.0/users/<redact>/sendMail HTTP/1.1" 202 0
2019-11-10 18:56:01,520 - O365.connection:_internal_request:726 - INFO - Received response (202) from URL https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:01,521 - crm_email.crm_email:message:76 - INFO - Email sent to Recipients count: 1
**2019-11-10 18:56:02,807 - O365.connection:_fs_token_lock:679 - DEBUG - Waking up and rechecking token file for update from other instance...**
**2019-11-10 18:56:02,808 - O365.connection:_fs_token_lock:684 - INFO - Token was refreshed by another instance...using that one**
2019-11-10 18:56:02,808 - O365.connection:_internal_request:721 - INFO - Requesting (POST) URL: https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:02,808 - O365.connection:_internal_request:722 - INFO - Request parameters: {'params'}
2019-11-10 18:56:02,808 - requests_oauthlib.oauth2_session:request:386 - DEBUG - Invoking 0 protected resource request hooks.
2019-11-10 18:56:02,808 - requests_oauthlib.oauth2_session:request:392 - DEBUG - Adding token {'token'} to request.
2019-11-10 18:56:02,809 - requests_oauthlib.oauth2_session:request:421 - DEBUG - Requesting url https://graph.microsoft.com/v1.0/users/<redact>/sendMail using method post.
2019-11-10 18:56:02,809 - requests_oauthlib.oauth2_session:request:422 - DEBUG - Supplying headers {'headers'} and data {'data'}
2019-11-10 18:56:02,809 - requests_oauthlib.oauth2_session:request:423 - DEBUG - Passing through key word arguments {}.
2019-11-10 18:56:02,815 - urllib3.connectionpool:_new_conn:955 - DEBUG - Starting new HTTPS connection (1): graph.microsoft.com:443
2019-11-10 18:56:03,251 - urllib3.connectionpool:_make_request:428 - DEBUG - https://graph.microsoft.com:443 "POST /v1.0/users/<redact>/sendMail HTTP/1.1" 202 0
2019-11-10 18:56:03,252 - O365.connection:_internal_request:726 - INFO - Received response (202) from URL https://graph.microsoft.com/v1.0/users/<redact>/sendMail
2019-11-10 18:56:03,252 - crm_email.crm_email:message:76 - INFO - Email sent to Recipients count: 1
```

**Comment by janscas** [comment]
Ok, I've reviewed the PR. 
I write here so everything stays in the same place although I reviewed some lines in the PR.

Here are muy comments:

1.  In my implementation of `should_refresh_token` I forgot to change the session token. I'll add a commit that will update the session token when the refresh was done by another token (simply: `self.session.token = self.token_backend.token`).

1. I think the connection shouldn't know anything about the token backend. Simply image that at the end the connection will have to introspect every case of posible token backends to take the proper action. This can be a large list of `if isinstance(self.token_backend, XXXX)`. Instead what I like is to let each token_backend decide what they need to do.

    So I think what we can do to change as little as posible from the connection is to **pass the connection itself to the `should_refresh_token` method**. So the new method signature will be:

    `def should_refresh_token(self, con=None):`

    Then the `_internal_request` will always pass 'self` (the connection) into this method.
    With this change in place you can move all the logic into the token backend.


1. I don't want to introduce new dependencies into the library. I don't want all the users to install portalocker when they maybe don't need that. In fact it's a pretty special case. So, first, the api must change to allow the modifications you did, but without introducing a new dependency. Therefore the FileSystemTokenBackend will stay the same (the fs_wait property can stay but I think it's not needed now). Then a new LockedFileSystemTokenBackend (for example) can be subclassed. The code for this subclass can stay in the token_backends.py in the examples directory (as I did with the Firestore example). The `_fs_token_lock` code can be integrated into the `should_refresh_token` code method of this new subclass.

1. I like the new Token properties. I've commented some minor thinks on the PR

**Comment by riptusk331** [comment]
I had a feeling we probably wouldn't want to modify `Connection` too much - I agree with your sentiments of trying to keep all this logic on the backend. Good idea on passing the connection to the backend as well. Also agree that it does not make sense to add new dependencies (portalocker) unless you really need it in this special case. I'll modify based on your suggestions into a new `LockedFileSystemTokenBackend` sublcass and amend the PR.

**Comment by riptusk331** [comment]
Pushed the discussed token.py changes to #356 ... but I see you added a commit (b1d2ac5) making the change to `should_refresh_token` to accept a `Connection`. I pushed my commit before I had a chance to pull yours in, so you may just have to merge...changes do not conflict.

**Comment by riptusk331** [comment]
See #357 for subclassed lockable file back end example. This now puts all the locking logic on the backend...the `Connection` class knows none the better (kind of). I've tested and it works nicely. I'll start using this in my company environment which will provide more robust testing....and see how it goes.

**Comment by janscas** [comment]
Great!! Let me know if it works as expected!

Thank you very much

**Comment by riptusk331** [comment]
I am going to close this issue as I believe we've solved it. Very nice to work with you on this. I'll report back with my testing.

How often do you push releases to PyPI?

**Comment by janscas** [comment]
> How often do you push releases to PyPI?

Every month or so....

**Comment by riptusk331** [comment]
sadly the issue 503 issue still persists in the production environment. even with the file locking and single request of a token. i'm determined to get to the bottom of this, so i installed [mitmproxy](https://mitmproxy.org/) to sniff all the http traffic in and out of the container. so far i've only ran it on my laptop, but i plan to deploy it in the production environment (where the issue is occuring) in the next day or 2.

at least on my laptop (where everything is succesful and i don't get a 503 error), i've verified that only one token refresh POST request is being made, followed by individual POST requests to the sendMail endpoint for each e-mail. there are no duplicate requests...and the content of the requests look fine. so the application is functioning as intended (meaning we are not sending duplicate token refresh requests with concurrent instances anymore). so i am feeling pretty confident that this is something on microsoft's end, unless there is something weird going on in my production environment. based on my logs though, i don't think so.

i've gotten in touch with someone from the graph team at microsoft and relayed this information, and they were inclined to agree that something strange is going on with the service. they asked me to send them the request/response headers & body. once i get mitmproxy up in my prod environment, i'll do so.

i will let you know what the outcome is.

**Comment by janscas** [comment]
Strange! I've reviewed the code and it seems there's nothing wrong to me.

Well let me know if you have any news. I'll reopen this.

**Comment by riptusk331** [comment]
Have been testing this for over a week and the issue has not occurred since I reached out to MS. Has been functioning perfectly & I get 202 responses on everything. File blocking is working as desired with the new backend & methods.

So I can say with a decent amount of confidence that this was something strange happening at MS...especially since the e-mails were still sending. Up to you if you want to keep the issue open...I will check back in with any update I have.

**Comment by janscas** [comment]
Greant news! 
I'll close this as it seems solved.
